Reduce bin size by excluding unneeded parts

clumsy-stefan opened the pull request:
Trying to reduce size of the compiled binaries by  excluding certain parts of the code if not needed.

See definitions at the end of  'define_plugin_sets.h'
 Currently done:
- Casting all controllers that are not included
- Casting all MQTT related code
- Casting all Domoticz related code

clumsy-stefan committed:
Reduce bin size by excluding unneded parts

Trying to reduce size of the compiled binaries by  excluding certain parts of the code if not needed.

See definitions at the end of  'define_plugin_sets.h'
 Currently done:
- Casting all controllers that are not included
- Casting all MQTT related code
- Casting all Domoticz related code

clumsy-stefan committed:
Reduce bin size -- Typos.

Corrected some typos...

TD-er commented on src/define_plugin_sets.h:
```diff
@@ -814,10 +814,10 @@ To create/register a plugin, you have to :
 #endif
 
-#if defined(USES_C082) || defined (USES_C005) || defined(USES_C006) || defined(USES_C014)
+#if defined(USES_C001) || defined(USES_C002) || defined (USES_C005) || defined(USES_C006) || defined(USES_C014)
```
C001 does not use MQTT, but it does share some code with C002 since they both are made for Domoticz.

clumsy-stefan replied:
ok, will change it back..

TD-er commented:
Maybe also wrap the include of PubSubClient with this `USES_MQTT`

jimmys01 commented:
Cool idea. 
Does the linker automaticaly exclude code that is not referenced anyway? At least for languages like Java I know it does.

TD-er commented:
> Cool idea.
> Does the linker automaticaly exclude code that is not referenced anyway? At least for languages like Java I know it does.

Well that was my assumption, but it looks like @clumsy-stefan did experience otherwise.
The linker can exclude object files not being used, but if (unused) code is in the object file, it will be hard to get it out of it during linking.
I know there are flags to determine what should be combined in an object file and what should have its own.
This can be set at various levels, like namespaces, classes or even per function.
An object file per function does take a lot more time to build, and in some cases it will not lead to smaller binaries since object files each have their own offset tables included.

If a function does refer to two completely different classes which are mutually exclusive, then the linker cannot split them apart and will include both.
For example let's assume we would be using HTTP or HTTPS and have a boolean parameter to tell us which one to use. Then both HTTP and HTTPS libraries will be included.
If on the other hand we split them into 2 separate functions and let the function name determine which one to use, then the linker can split them.  (but not if it is creating objects at class level and both are in the same class)

clumsy-stefan commented:
I don't understand why travis is complaining about:
```
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:940:63: error: 'C011_queue_element' does not name a type
11750
```
but only for C011, C012, C015, C016, C017 and C018...

can someone enlighten me?

TD-er commented:
> I don't understand why travis is complaining about:
> 
> ```
> /home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:940:63: error: 'C011_queue_element' does not name a type
> 11750
> ```
> 
> but only for C011, C012, C015, C016, C017 and C018...
> 
> can someone enlighten me?

I've seen something similar when trying this for C018 last week.

uzi18 commented:
we should try with platformio 3.6.7, first
maybe something changed in build system

TD-er commented:
I can clear the Travis cache, just to be sure?

TD-er commented:
I did clear the caches and restarted the build.
So please do not add a commit to this PR until that one has finished building.

clumsy-stefan commented:
I had a quick look through the code but I have absolutel no clue why this happens... and why only for these controllers... probably this has some relations to the macro-definition in `c_plugin_helper_h` as it happens only with controllers that have `DEFINE_Cxxx_DELAY_QUEUE_MACRO` defined and have 2 digits in the name...
as far as I could find out until now...

TD-er commented:
Hmm the 2 digits may be something.
If a digit sequence has a single leading 0, it may be considered octal.

TD-er commented:
But why does it not have an issue with C010 then?

clumsy-stefan commented:
> But why does it not have an issue with C010 then?

good question... actually I anyway have no idea how macro's work ;)

TD-er commented:
Well I have StackOverflow, so I can program macros :) (as long as SO is not offline....)

I think I have to change that macro to split the NNN and M parameter.
Right now, I have to use 2 since I have no clue how to glue leading zeroes in macros to a fixed length.

But maybe I should change
```
DEFINE_Cxxx_DELAY_QUEUE_MACRO(001,  1)
DEFINE_Cxxx_DELAY_QUEUE_MACRO(011,  11)
```

into this:
```
DEFINE_Cxxx_DELAY_QUEUE_MACRO(00,  1)
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0,  11)
```

And of course also the macro itself

TD-er commented:
I think if we change `##NNN##` in that macro into `##NNN####M##` then we can use the proposed change in the define call.

clumsy-stefan commented:
whatever it takes... I won't be of much help in the macro-sphere ;)

TD-er commented:
Yep, if I change it to this, I am able to place the queue element struct for C017 into the  `#ifdef USES_C017` wrappers

```c++

// This macro defines the code needed to create the 'process_c##NNN####M##_delay_queue()'
// function and all needed objects and forward declarations.
// It is a macro to prevent common typo errors.
// This function will perform the (re)scheduling and mark if it is processed (and can be removed)
// The controller itself must implement the 'do_process_c004_delay_queue' function to actually
// send the data.
// Its return value must state whether it can be marked 'Processed'.
// N.B. some controllers only can send one value per iteration, so a returned "false" can mean it
//      was still successful. The controller should keep track of the last value sent
//      in the element stored in the queue.
#define DEFINE_Cxxx_DELAY_QUEUE_MACRO(NNN, M)                                                                        \
  ControllerDelayHandlerStruct<C##NNN####M##_queue_element>C##NNN####M##_DelayHandler;                               \
  bool do_process_c##NNN####M##_delay_queue(int controller_number,                                                   \
                                           const C##NNN####M##_queue_element & element,                              \
                                           ControllerSettingsStruct & ControllerSettings);                           \
  void process_c##NNN####M##_delay_queue() {                                                                         \
    C##NNN####M##_queue_element *element(C##NNN####M##_DelayHandler.getNext());                                      \
    if (element == NULL) return;                                                                                     \
    MakeControllerSettings (ControllerSettings);                                                                     \
    LoadControllerSettings(element->controller_idx, ControllerSettings);                                             \
    C##NNN####M##_DelayHandler.configureControllerSettings(ControllerSettings);                                      \
    if (!WiFiConnected(10)) {                                                                                       \
      scheduleNextDelayQueue(TIMER_C##NNN####M##_DELAY_QUEUE, C##NNN####M##_DelayHandler.getNextScheduleTime());     \
      return;                                                                                              \
    }                                                                                                      \
    START_TIMER;                                                                                           \
    C##NNN####M##_DelayHandler.markProcessed(do_process_c##NNN####M##_delay_queue(M, *element, ControllerSettings)); \
    STOP_TIMER(C##NNN####M##_DELAY_QUEUE);                                                                      \
    scheduleNextDelayQueue(TIMER_C##NNN####M##_DELAY_QUEUE, C##NNN####M##_DelayHandler.getNextScheduleTime());       \
  }

// Define the function wrappers to handle the calling to Cxxx_DelayHandler etc.
// If someone knows how to add leading zeros in macros, please be my guest :)
#ifdef USES_C001
DEFINE_Cxxx_DELAY_QUEUE_MACRO(00,  1)
#endif // ifdef USES_C001
#ifdef USES_C003
DEFINE_Cxxx_DELAY_QUEUE_MACRO(00,  3)
#endif // ifdef USES_C003
#ifdef USES_C004
DEFINE_Cxxx_DELAY_QUEUE_MACRO(00,  4)
#endif // ifdef USES_C004
#ifdef USES_C007
DEFINE_Cxxx_DELAY_QUEUE_MACRO(00,  7)
#endif // ifdef USES_C007
#ifdef USES_C008
DEFINE_Cxxx_DELAY_QUEUE_MACRO(00, 8)
#endif // ifdef USES_C008
#ifdef USES_C009
DEFINE_Cxxx_DELAY_QUEUE_MACRO(00, 9)
#endif // ifdef USES_C009
#ifdef USES_C010
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0,  10)
#endif // ifdef USES_C010
#ifdef USES_C011
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0,  11)
#endif // ifdef USES_C011
#ifdef USES_C012
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0,  12)
#endif // ifdef USES_C012

/*
 #ifdef USES_C013
   DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 13)
 #endif
 */

/*
 #ifdef USES_C014
   DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 14)
 #endif
 */
#ifdef USES_C015
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 15)
#endif // ifdef USES_C015

#ifdef USES_C016
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 16)
#endif // ifdef USES_C016


#ifdef USES_C017
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 17)
#endif // ifdef USES_C017

#ifdef USES_C018
DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 18)
#endif


/*
 #ifdef USES_C019
   DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 19)
 #endif
 */

/*
 #ifdef USES_C020
   DEFINE_Cxxx_DELAY_QUEUE_MACRO(0, 20)
 #endif
 */
```

uzi18 commented:
it. should. be possible. to resolve these macros into destination source with gcc

clumsy-stefan commented:
@TD-er can you add the change to this PR so we could test if it works then?

TD-er commented:
@clumsy-stefan You can just cut and paste this part, as I also changed all define calls.

clumsy-stefan commented:
ok, so I include it as a commit in my PR?!

TD-er commented:
> ok, so I include it as a commit in my PR?!

Yep

clumsy-stefan committed:
Reduce bin size -- changed macro definitions

Changed macro definitions (by @TD_er) to hopefully make travis happy again...

clumsy-stefan commented:
> > ok, so I include it as a commit in my PR?!
> 
> Yep

did not help 😞 

but my compiler is still happy with these changes...

EDIT:
PS: did someone mention how superior platformIO is?!?! 😁 😀

TD-er commented:
> PS: did someone mention how superior platformIO is?!?! 😁 😀

Nope, please enlighten me ;)

clumsy-stefan commented:
> Nope, please enlighten me ;)

different discussion with same topic ;) (https://github.com/letscontrolit/ESPEasy/commit/54117e908fd48771bc810718de7a1b13bde671ec)

TD-er commented:
I think we should leave those C0xx_queue_element classes in there and hope for the linker to remove them.
I really have no clue what's happening here.
It may have something to do at what stage of the building process the defines are set or unset.
For example if I set the defines in the pre_extra_script.py, then it does seem to build just fine, but the other defines we set in a .h file we include.

So maybe the preprocessor then already has defined some forward declarations.

clumsy-stefan commented:
> Maybe also wrap the include of PubSubClient with this `USES_MQTT`

did that, saves anoher 500 bytes roughly

clumsy-stefan committed:
Reduce bin size -- add PubSubClient to #ifdef's

Added all references to PubSubClient to the `#ifdef USES_MQTT`castings and included it only if MQTT is used or P37 is enabled.

clumsy-stefan commented:
> I think we should leave those C0xx_queue_element classes in there and hope for the linker to remove them.
> I really have no clue what's happening here.

agree with that...

> So maybe the preprocessor then already has defined some forward declarations.

but why then only the ones after C011? and not C010 and lower?

clumsy-stefan commented:
btw: did you see that the very first build (custom_ESP8266_4M) did go through fine?! Also strange...

TD-er commented:
> btw: did you see that the very first build (custom_ESP8266_4M) did go through fine?! Also strange...

Yep, thats the one using the pre_extra_script.py.
That one sets the defines right at the start of the compile run.
So it is already set as if it is given as a compiler command line argument.

jimmys01 commented:
[src.zip](https://github.com/letscontrolit/ESPEasy/files/3550425/src.zip)

This compiles with no issues.
Basically it is the code reverted to 322b833  and I changed the _CPlugin_Helper.h

In short 
```
#ifdef USES_C001
/*********************************************************************************************\
* C001_queue_element for queueing requests for C001.
\*********************************************************************************************/
#define C001_queue_element simple_queue_element_string_only
#endif //USES_C001
```

Changed to 


```
/*********************************************************************************************\
* C001_queue_element for queueing requests for C001.
\*********************************************************************************************/
#define C001_queue_element simple_queue_element_string_only

```

clumsy-stefan committed:
Reduce bin size -- cast blynk code

save some bytes by casting unused blynk code

clumsy-stefan committed:
Revert "Reduce bin size -- cast blynk code"

This reverts commit 1050a07f788d155de20613638b79ad60d3ac2ecc.

TD-er commented:
@jimmys01 The whole idea was to exclude as much as possible to minimize the build size.
And it still is not yet (entirely) clear to me why those C0xx_queue_element classes cannot be removed from the source by just using the #ifdef wrappers.

Until that's clear, we may be left with binary builds which are bigger than they need to be.
But for now, we should revert the changes in _CPlugin_Helper.h just to be able to go on.

Problem is, if we change the builds to use a selection based on Python scripts alone, we completely cut off builds made by the ArduinoIDE and it seems there are still users building like that :)

jimmys01 commented:
> But for now, we should revert the changes in _CPlugin_Helper.h just to be able to go on.

exactly my thoughts

clumsy-stefan committed:
Reduce bin size -- enable all controller structs again

enable all controller structs to make platformIO happy again (hopefully)..
However, this does not seem to add any  size at least not on Arduino IDE....

clumsy-stefan commented:
> But for now, we should revert the changes in _CPlugin_Helper.h just to be able to go on.

that's what I did with commit https://github.com/letscontrolit/ESPEasy/pull/2575/commits/679ca3e4a1901db9ca16d7bb4c57bb95a0776645

ok, so now I broke something completely, but can't see where and how... to me it looks a bit liek an unmatched/ended #ifdef/#endif pair...but can't find where...

PS: My ArduinoIDE is still building fine....

uzi18 commented:
@TD-er there is maybe some kind of solution - on STM32 core there is build_opt.h file you can declare whatever you need for all files build with Arduino IDE, not sure if it is supported in esp8266 core

https://github.com/stm32duino/wiki/wiki/Custom-definitions

https://github.com/stm32duino/Arduino_Core_STM32/commit/130c5ab3ba52d282a86cba2774481da9222b2484

https://github.com/stm32duino/Arduino_Core_STM32/blob/026eb42ddbf3284e74ac98d887731eda0cfc7926/platform.txt#L89-L97

jimmys01 commented:
@ clumsy-stefan #2576  compiled.
 it is ther .zip I posted above.

TD-er commented:
@uzi18 I was also looking into splitting the uses defines into several .h files and they can be included as build parameters. That should be similar to what the pre_extra_scrip.py does.

uzi18 commented:
@TD-er with this idea it will be supported with arduino ide so we can move to more compatible mode for @clumsy-stefan 
one can add there in file -Idefaults.h or -DUSES_C001

clumsy-stefan commented:
@jimmys01 I had a loko at _CPlugin_Helper.h in your zip, and it^s essentially the same than I have posted in tha last commit (https://github.com/letscontrolit/ESPEasy/pull/2575/commits/679ca3e4a1901db9ca16d7bb4c57bb95a0776645). Please have a look. I have no clue why it doesn't compile here... It works smoothly in Arduine...I'm sure I have a typo somewhere...

uzi18 commented:
@clumsy-stefan it does not build on ESP32 platform

clumsy-stefan commented:
> @clumsy-stefan it does not build on ESP32 platform

only not on ESP32? never tested that actually.... but thanks for the pointer...

uzi18 commented:
find line "Compiling .pio/build/esp-wrover-kit_test_1M8_partition/FrameworkArduino/FunctionalInterrupt.cpp.o"

https://travis-ci.org/letscontrolit/ESPEasy/builds/577872460?utm_source=github_status&utm_medium=notification

jimmys01 commented:
here is the diff between your latest commit here and the .zip
https://github.com/jimmys01/ESPEasy/commit/8a6a43b21c5ba316292a125a230d886c70817004

clumsy-stefan commented:
> here is the diff between your latest commit here and the .zip
> [jimmys01@8a6a43b](https://github.com/jimmys01/ESPEasy/commit/8a6a43b21c5ba316292a125a230d886c70817004)

Hmm... but _CPluigin_Helper.h should also have differences?!

jimmys01 commented:
yea. i forgot to add this file to the commit. Anyway, I also dont know why it doesnt compile.

clumsy-stefan commented:
@jimmys01 thanks for your help! I mean the changes in this diff are quite overseeable and not really complex, so I don't see any reason why it shouldn' compile with this version? Really strange... like with the controller defines that don't work on PIO....

clumsy-stefan commented:
@uzi18 just compiled a ESP32 version here on my ArduinoIDE whitch ran through fine...

I'm sure I'm missing something here, but I have no clue anymore where to look for...

jimmys01 commented:
in my opinion revert to my .zip and then add small stuff  to see what breaks things. Or install Visual Studio Code with platformio and change your life forever :-p

clumsy-stefan commented:
@jimmys01 😁 my life is fine... my versions compile fine with ArduinoIDE and save me quite a lot of memory... so I can live fine with that... but I thought you PIO guys also want to profit from the improved memory 😛

clumsy-stefan commented:
my problem currently is, that it compiles without any warnings or errors on ArduinoIDE, so I can't really test why it doesn't compiel ond PIO.... 
I did a number of smaller commits exactly because of that, but it somehow didn^t help....

but any help apreciated...

uzi18 commented:
@clumsy-stefan you can split it into smaller parts/hunks and add one by one to find what couse problem

clumsy-stefan commented:
I think I found it... It seems that commit https://github.com/letscontrolit/ESPEasy/pull/2575/commits/1050a07f788d155de20613638b79ad60d3ac2ecc which is excluding the blynk code caused the issues...

so the remaining question is still why the controller plugins in _CPlugin_Helper.h can't be excluded... but this can be solved in a later stage, 

let's see if travis is happy now. if so  @TD-er can include this PR as as start... should reduce the builds by 50-100k...

and thanks for everybody's help!! really appreciated

jimmys01 commented:
As far as I have tested there is no significant difference in the size of the builds before this PR.
Perhaps Arduino IDE compiles stuff so much diferently?

uzi18 commented:
@clumsy-stefan try add #include "define_plugin_sets.h" in _CPlugin_Helper.h

clumsy-stefan commented:
Compiling with version from mega-20190827 (#4a8cfd2d9b1fbfea27fa7b7829cb85a7ffc29301):
```849840 Aug 28 22:16 ESPEasy.ino.d1_mini.bin```
using this PR:
```841472 Aug 28 22:20 ESPEasy.ino.d1_mini.bin```
So roughly 8Kb (just realise I had a 0 too much in my previous post, meant 5-10k), Probably not a lot, but when it's tight anyways....

Also probably the main difference is, that I don't use any MQTT controller, so that code is irrelevant to me and overall the project, so it could be, that this is what makes a difference to my builds!

Also the idea was to exclude the controller stuff, but this is not (yet) working as discussed above...

TD-er commented:
> Also the idea was to exclude the controller stuff, but this is not (yet) working as discussed above...

I don't think that controller part stuff is going to make a lot of difference.
It is not including lots of other code and only the templated code is going to add some to the binary size.
